File: clients/java/src/main/java/io/zeebe/client/impl/subscription/JobPoller.java
/*
 * Copyright © 2017 camunda services GmbH (info@camunda.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.zeebe.client.impl.subscription;

import io.grpc.stub.StreamObserver;
import io.zeebe.client.api.response.ActivatedJob;
import io.zeebe.client.impl.Loggers;
import io.zeebe.client.impl.ZeebeObjectMapper;
import io.zeebe.client.impl.response.ActivatedJobImpl;
import io.zeebe.gateway.protocol.GatewayGrpc.GatewayStub;
import io.zeebe.gateway.protocol.GatewayOuterClass.ActivateJobsRequest.Builder;
import io.zeebe.gateway.protocol.GatewayOuterClass.ActivateJobsResponse;
import java.util.function.Consumer;
import org.slf4j.Logger;

public class JobPoller implements StreamObserver<ActivateJobsResponse> {

  private static final Logger LOG = Loggers.JOB_POLLER_LOGGER;

  private final GatewayStub gatewayStub;
  private final Builder requestBuilder;
  private final ZeebeObjectMapper objectMapper;

  private Consumer<ActivatedJob> jobConsumer;
  private Consumer<Integer> doneCallback;
  private int activatedJobs;

  public JobPoller(
      GatewayStub gatewayStub, Builder requestBuilder, ZeebeObjectMapper objectMapper) {
    this.gatewayStub = gatewayStub;
    this.requestBuilder = requestBuilder;
    this.objectMapper = objectMapper;
  }

  private void reset() {
    activatedJobs = 0;
  }

  void poll(
      int maxJobsToActivate, Consumer<ActivatedJob> jobConsumer, Consumer<Integer> doneCallback) {
    reset();

    requestBuilder.setMaxJobsToActivate(maxJobsToActivate);
    this.jobConsumer = jobConsumer;
    this.doneCallback = doneCallback;

    poll();
  }

  private void poll() {
    LOG.trace(
        "Polling at max {} jobs for worker {} and job type {}",
        requestBuilder.getMaxJobsToActivate(),
        requestBuilder.getWorker(),
        requestBuilder.getType());
    gatewayStub.activateJobs(requestBuilder.build(), this);
  }

  @Override
  public void onNext(ActivateJobsResponse activateJobsResponse) {
    activatedJobs += activateJobsResponse.getJobsCount();
    activateJobsResponse.getJobsList().stream()
        .map(job -> new ActivatedJobImpl(objectMapper, job))
        .forEach(jobConsumer);
  }

  @Override
  public void onError(Throwable throwable) {
    LOG.warn(
        "Failed to activated jobs for worker {} and job type {}",
        requestBuilder.getWorker(),
        requestBuilder.getType(),
        throwable);
    pollingDone();
  }

  @Override
  public void onCompleted() {
    pollingDone();
  }

  private void pollingDone() {
    if (activatedJobs > 0) {
      LOG.info(
          "Activated {} jobs for worker {} and job type {}",
          activatedJobs,
          requestBuilder.getWorker(),
          requestBuilder.getType());
    } else {
      LOG.trace(
          "No jobs activated for worker {} and job type {}",
          requestBuilder.getWorker(),
          requestBuilder.getType());
    }
    doneCallback.accept(activatedJobs);
  }
}

File: clients/go/pkg/worker/jobPoller.go
// Copyright © 2018 Camunda Services GmbH (info@camunda.com)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package worker

import (
	"context"
	"fmt"
	"github.com/camunda/zeebe/clients/go/pkg/entities"
	"github.com/camunda/zeebe/clients/go/pkg/pb"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"io"
	"log"
	"sync"
	"time"
)

type jobPoller struct {
	client         pb.GatewayClient
	request        *pb.ActivateJobsRequest
	requestTimeout time.Duration
	maxJobsActive  int
	pollInterval   time.Duration

	jobQueue       chan entities.Job
	workerFinished chan bool
	closeSignal    chan struct{}
	remaining      int
	threshold      int
	metrics        JobWorkerMetrics
	shouldRetry    func(context.Context, error) bool
}

func (poller *jobPoller) poll(closeWait *sync.WaitGroup) {
	defer closeWait.Done()

	// initial poll
	poller.activateJobs()

	for {
		select {
		// either a job was finished
		case <-poller.workerFinished:
			poller.remaining--
			poller.setJobsRemainingCountMetric(poller.remaining)
		// or the poll interval exceeded
		case <-time.After(poller.pollInterval):
		// or poller should stop
		case <-poller.closeSignal:
			poller.setJobsRemainingCountMetric(0)
			return
		}

		if poller.shouldActivateJobs() {
			poller.activateJobs()
		}
	}
}

func (poller *jobPoller) shouldActivateJobs() bool {
	return poller.remaining <= poller.threshold
}

func (poller *jobPoller) activateJobs() {
	ctx, cancel := context.WithTimeout(context.Background(), poller.requestTimeout)
	defer cancel()

	poller.request.MaxJobsToActivate = int32(poller.maxJobsActive - poller.remaining)
	stream, err := poller.openStream(ctx)
	if err != nil {
		log.Println(err.Error())
		return
	}

	for {
		response, err := stream.Recv()
		if err != nil {
			if poller.shouldRetry(ctx, err) {
				// the headers are outdated and need to be rebuilt
				stream, err = poller.openStream(ctx)
				if err != nil {
					log.Printf("Failed to reopen job polling stream: %v\n", err)
					break
				}
				continue
			}

			if err != io.EOF && status.Code(err) != codes.ResourceExhausted {
				log.Printf("Failed to activate jobs for worker '%s': %v\n", poller.request.Worker, err)
			}

			break
		}

		poller.remaining += len(response.Jobs)
		poller.setJobsRemainingCountMetric(poller.remaining)
		for _, job := range response.Jobs {
			poller.jobQueue <- entities.Job{ActivatedJob: job}
		}
	}
}

func (poller *jobPoller) openStream(ctx context.Context) (pb.Gateway_ActivateJobsClient, error) {
	stream, err := poller.client.ActivateJobs(ctx, poller.request)
	if err != nil {
		if poller.shouldRetry(ctx, err) {
			return poller.openStream(ctx)
		}
		return nil, fmt.Errorf("worker '%s' failed to open job stream: %w", poller.request.Worker, err)
	}

	return stream, nil
}

func (poller *jobPoller) setJobsRemainingCountMetric(count int) {
	if poller.metrics != nil {
		poller.metrics.SetJobsRemainingCount(poller.request.GetType(), count)
	}
}
